hydromt_wflow/wflow.py

```diff
@@ -3147,37 +3147,48 @@ def setup_cold_states(
             self.set_config(option, states_config[option])
 
     # I/O
-    def read(self):
+    def read(
+        self,
+        config_fn: Union[Path, str] = None,
+        geom_fn: Union[Path, str] = "staticgeoms",
+    ):
         """Read the complete model schematization and configuration from file."""
-        self.read_config()
+        self.read_config(config_fn=config_fn)
         self.read_grid()
+        self.read_geoms(geom_fn=geom_fn)
+        self.read_forcing()
         self.read_intbl()
         self.read_tables()
-        self.read_geoms()
         self.read_states()
-        self.read_forcing()
         self.logger.info("Model read")
 
-    def write(self):
+    def write(
+        self,
+        config_fn: str = None,
+        grid_fn: Union[Path, str] = "staticmaps.nc",
+        geom_fn: Union[Path, str] = "staticgeoms",
+        forcing_fn: Union[Path, str] = None,
+    ):
         """Write the complete model schematization and configuration to file."""
         self.logger.info(f"Write model data to {self.root}")
         # if in r, r+ mode, only write updated components
         if not self._write:
             self.logger.warning("Cannot write in read-only mode")
             return
         self.write_data_catalog()
-        if self.config:  # try to read default if not yet set
```

Comment:
Most filenames are read from the config and as the comment says, we did find some cases where at that stage config was not yet read. To be sure, I would then leave a call to self.config to be sure it is read. You can keep the writing as last.

Reply:
Agreed!